expression/builtin_math.go

```diff
@@ -1011,12 +1020,16 @@ func (b *builtinRandWithSeedSig) evalReal(row chunk.Row) (float64, bool, error)
 		return 0, true, err
 	}
 	if b.randGen == nil {
-		if isNull {
-			// When seed is NULL, it is equal to RAND().
-			b.randGen = rand.New(rand.NewSource(time.Now().UnixNano()))
-		} else {
-			b.randGen = rand.New(rand.NewSource(seed))
+		b.mu.Lock()
+		if b.randGen == nil {
+			if isNull {
+				// When seed is NULL, it is equal to RAND().
+				b.randGen = rand.New(rand.NewSource(time.Now().UnixNano()))
+			} else {
+				b.randGen = rand.New(rand.NewSource(seed))
```

Comment:
Something not related to this pr, according to mysql, `with a nonconstant initializer argument (such as a column name), the seed is initialized with the value for each invocation of RAND().`, so is it right to only init it once?

Reply:
You're right. The implementation of `rand(N)` in TiDB is not compatible with MySQL.